### ISSUE TITLE
make subtyping explicit post analysis

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#112651]
We should add a MIR pass after analysis (and before any optimizations) which makes subtyping explicit in the MIR. This is a robust way to fix https://github.com/rust-lang/rust/issues/107205, preventing similar issues in the future. I don't quite know how to do that, probably by adding an explicit `ProjectionElem::Cast` (probably reusing `OpaqueCast` for that).

**Comment by lcnr** [comment]
When implementing this, also handle all the `FIXME(#112651)` added in #112307

**Comment by goffrie** [comment]
Should this be closed now that #115025 has been merged?

**Comment by lcnr** [comment]
yes, this is totally fine and even expected. However, we wanted to make subtyping explicit to prevent bugs like #107205 going forward:

While subtyping does not change the layout of a type, it does change trait selection. This broke stuff in two ways:

Optimizations, e.g. const prop, but also inlining, can change
```rust
let x: subtype = whatever;
let y: supertype = x;
y.requires_trait_selection(); // e.g. by casting it to a trait object
```
to
```rust
let x: subtype;
x.requires_trait_selection();
```

We also currently use typed values both in the MIR, e.g. https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/consts/enum.Const.html#variant.Val and during codegen https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/mir/operand/struct.OperandRef.html (this uses `TyAndLayout`). If we propagate that value to a place which has a super type but don't update its type, using the type specified as part of the *value* is wrong.

> That subtyping only arises dynamically; as far as the type system is concerned, we don't know what `y` points to so the fact that the underlying value has type `fn(fn(&'static ()))` should not be an issue?

It can be known by optimizations/ctfe, which can be unsound if they accidentally use the type of the value passed into the function to do trait selection.

----

I think ideally we would not have values which are typed, especially not in codegen, where the actual unsoundness happened. However, it was very difficult to change `OperandRef` to only use a `Layout` so I did not pursue that further.

**Comment by RalfJung** [comment]
> While subtyping does not change the layout of a type, it does change trait selection.

Uh, okay, that sounds bad.

> It can be known by optimizations/ctfe, which can be unsound if they accidentally use the type of the value passed into the function to do trait selection.

CTFE considers function argument passing to be a transmute. The caller and callee might have completely different but ABI-compatible types. Many things would go wrong loudly if we used the caller type ever inside the callee.

Also, in CTFE values are only ephemerally typed, during the execution of a single machine step (`ImmTy`/`OpTy`). The values that persist between machine steps are stored in `Immediate` or in memory, and are hence not typed. This matches the (envisioned) operational semantics where the untyped raw representation gets turned into a structured and typed [`Value`](https://github.com/minirust/minirust/blob/master/spec/lang/values.md) when the AM needs to act on it (e.g. to add two integers together, or to dereference a pointer), and then back into an untyped raw representation when it is e.g. stored inside some place.

The only optimization where type information from the caller could leak to the callee is inlining, I think?

**Comment by RalfJung** [comment]
> alternatively we could try to change any op which relies on type info to store the type info by itself, without relying on the type of the affected local?

yeah that'd be like wasm a bit, having each operation carry the types it needs and basically making locals untyped. Not sure if that would be a good fit for MIR though, it seems highly redundant.

As you say, codegen just uses a bad structure here. Though you also had to fix some optimizations and I am less sure what you be done structurally to help them avoid this pitfall...

But overall if we say that inlining is a transmutation site then the explicit projections you had planned should work, right? And inlining has to be a transmutation site in general when function pointers are involved because we allow some type mismatches as long as things are ABI-compatible.

> though ofc now you have to remember to always explicitly transmute when looking into nested functions, which is also easy to mess up

What do you mean? We don't "look into" nested functions, do we? We just paste the MIR of the callee into the caller. And then if this is a call through a function pointer we need to add transmute. We already need to do that, the caller might use type `NonZeroU32` and the callee `u32` and obviously things explode if we don't insert some explicit transmute.

When the call is a direct call of a fn item, then I don't think there is a possible issue here?

**Comment by RalfJung** [comment]
> Quickly looking over the MIR, the typed operations are Rvalue::Cast with CastKind::Unsize and TerminatorKind::Call (the type of func is used to fetch the right instance).

Now I am confused. I thought we'd need types basically everywhere a local shows up. Why only these two, what is special about them? The type of a local is relevant in a lot more places after all. That doesn't sound so bad then.

> requires inlining of function pointers to explicitly transmute the args if types differ (that happens anyways)

Right, inlining of function pointers without transmute is already wrong, so this changes nothing here.

I wonder what other MIR consumers think. Cc @rust-lang/wg-mir-opt

**Comment by lcnr** [comment]
> Now I am confused. I thought we'd need types basically everywhere a local shows up. Why only these two, what is special about them? The type of a local is relevant in a lot more places after all. That doesn't sound so bad then.

subtyping can't change layout, so it's only when interacting with the trait system that it makes a difference. Most MIR ops don't call the trait solver, so subtyping doesn't impact them

**Comment by BoxyUwU** [comment]
The only time that subtyping can change trait solver results is higher ranked subtyping (e.g. `for<'a> fn(&'a u32)` -> `fn(&'static u32)`), does this imply that if we started using coercions for such operations (breaking change :3) this issue would become moot as eliding a local and using a supertype would not be observable?

**Comment by lcnr** [comment]
> The only time that subtyping can change trait solver results is higher ranked subtyping (e.g. `for<'a> fn(&'a u32)` -> `fn(&'static u32)`), does this imply that if we started using coercions for such operations (breaking change :3) this issue would become moot as eliding a local and using a supertype would not be observable?

jup ✨
